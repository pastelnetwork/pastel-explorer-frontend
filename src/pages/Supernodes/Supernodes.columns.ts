export const SUPERNODE_IP_KEY = 'ip';

export const columns = [
  {
    width: 80,
    flexGrow: 1,
    label: 'pages.supernodes.supernodeIP',
    dataKey: SUPERNODE_IP_KEY,
    disableSort: false,
    className: 'col-supernode-ip',
    dataTitle: 'pages.supernodes.supernodeIP',
  },
];
