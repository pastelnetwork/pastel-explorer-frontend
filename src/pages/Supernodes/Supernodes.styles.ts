import styled from 'styled-components';
import { Grid } from '@material-ui/core';
import { CSVLink } from 'react-csv';

export const GridWrapper = styled(Grid)`
  .col-address {
    & > .MuiTableCell-root {
      padding-left: 0;
    }
  }

  .bold {
    font-weight: 700;
  }

  .title {
    font-size: 15px;
  }

  .copy-icon {
    margin-left: 0;
  }

  .ReactVirtualized__Table__headerRow {
    display: none;
  }

  .supernode-table {
    padding-left: 0;
    padding-right: 0;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }

  ${props => props.theme.breakpoints.down(1024)} {
    .address-link {
      max-width: calc(100vw - 185px);
    }

    .ReactVirtualized__Grid {
      min-width: unset;
    }

    .ReactVirtualized__Table__headerRow {
      display: none;
    }

    .ReactVirtualized__Table__row {
      flex-direction: column;
    }

    .ReactVirtualized__Table__rowColumn {
      display: flex;
      align-items: center;
      flex: unset !important;
      width: 100%;
      padding-left: 10px;
      padding-right: 10px;

      .hash-link {
        max-width: 68%;
      }

      .MuiTableCell-root {
        padding-left: 0;
        border-bottom: 0;
      }
    }
  }

  .ticket-title-wrapper {
    h4 {
      margin: 0;
      padding: 0;
    }
  }

  ${props => props.theme.breakpoints.down('xs')} {
    .ReactVirtualized__Table__rowColumn {
      &:before {
        width: 90px;
        min-width: 90px;
      }
    }
  }
`;

export const BlockWrapper = styled.div`
  margin-top: 24px;
  box-shadow: 0px 5px 6px rgb(16 16 16 / 6%);
  background: ${props => props.theme.sidebar.menu.background};
  border-radius: 10px;
  overflow: hidden;

  .table-title {
    margin: 0;
    padding: 18px 16px;
    background: ${props => props.theme.card.titleColor};
  }

  ${props => props.theme.breakpoints.down(800)} {
    .ticket-title-wrapper {
      h4 {
        margin-bottom: 5px;
      }
    }
  }

  ${props => props.theme.breakpoints.down(570)} {
    .filter-wrapper {
      display: flex;
      flex-direction: column;
    }

    .supernode-status {
      flex-direction: column;
    }
  }
`;

export const TitleWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  margin: 0;
  padding: 0;
  background: ${props => props.theme.card.titleColor};

  ${props => props.theme.breakpoints.down(800)} {
    flex-direction: column;
  }
`;

export const Title = styled.span`
  margin-right: 0;
`;

export const SubTitle = styled.span`
  font-size: 16px;
  font-weight: 400;

  ${props => props.theme.breakpoints.down('xs')} {
    font-size: 14px;
  }
`;

export const Status = styled.div`
  display: inline-flex;
  padding: 2px 10px;
  font-size: 14px;
  background: ${props => props.theme.supernodes.status.background.newStart};
  color: ${props => props.theme.supernodes.status.color};
  border-radius: 8px;

  &.enabled {
    background: ${props => props.theme.supernodes.status.background.enabled};
  }

  &.expired {
    background: ${props => props.theme.supernodes.status.background.expired};
  }
`;

export const TitleContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;

  ${props => props.theme.breakpoints.down('sm')} {
    flex-direction: column;
  }
`;

export const FilterBlock = styled.div`
  ${props => props.theme.breakpoints.down('sm')} {
    margin-top: 10px;
  }

  ${props => props.theme.breakpoints.down(570)} {
    text-align: center;

    .supernode-status {
      flex-direction: column;
      width: 100%;
    }
  }
`;

export const Gird = styled.div`
  display: flex;

  ${props => props.theme.breakpoints.down(1024)} {
    flex-wrap: wrap;
  }
`;

export const ExplorerMapColumn = styled.div`
  width: 60%;
  margin-right: 12px;

  ${props => props.theme.breakpoints.down(1024)} {
    width: 100%;
    margin-right: 0;
    margin-bottom: 12px;
  }
`;

export const SupernodeColumn = styled.div`
  width: calc(40% - 4px);

  ${props => props.theme.breakpoints.down(1024)} {
    width: 100%;
  }
`;

export const CSVLinkButton = styled(CSVLink)`
  margin-left: 10px;
  padding: 8px 25px;
  background: ${props => props.theme.sidebar.menu.toggle.switch};
  color: #fff;
  font-size: 16px;
  font-weight: 400;
  font-family: 'Gill Sans';
  cursor: pointer;
  border-radius: 10px;
  border: 0;
  outline: none;
  text-decoration: none;
  transition: all 0.5s ease;

  &:hover {
    background: ${props => props.theme.sidebar.menu.toggle.hover};
  }

  ${props => props.theme.breakpoints.down(570)} {
    display: inline-flex;
    margin-top: 10px;
    margin-left: auto;
    margin-right: auto;
  }

  @media screen and (max-width: 375px) {
    padding: 8px 16px;
    font-size: 15px;
  }
`;

export const ContentWrapper = styled.div`
  display: block;

  .masternode-detail {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr 1fr;
    gap: 10px;
  }

  .header-cell {
    display: inline-flex;
    align-items: center;
  }

  .nowrap {
    white-space: nowrap;
  }

  .title {
    font-weight: 400;
  }

  .MuiIconButton-root {
    .MuiSvgIcon-root {
      transition: all 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    }

    &.open {
      .MuiSvgIcon-root {
        transform: rotate(180deg);
      }
    }
  }

  .custom-table {
    .table__row {
      &.row_active {
        &,
        &:hover,
        &:nth-of-type(2n + 1) {
          background-color: ${props => props.theme.supernodes.table.even} !important;
        }
      }
    }

    .row_more_active {
      background-color: ${props => props.theme.supernodes.table.active};

      &:nth-of-type(2n + 1) {
        background-color: ${props => props.theme.supernodes.table.even} !important;
      }
    }

    .MuiSvgIcon-root {
      color: ${props => props.theme.palette.text.primary};
    }

    .wrapper-content {
      display: inline-flex;
      align-items: center;
    }

    .see-more {
      .MuiButtonBase-root {
        padding: 1px;
      }
    }
  }

  ${props => props.theme.breakpoints.down(1170)} {
    .custom-table {
      .table__row-header {
        display: none;
      }

      .table__row {
        display: flex;
        flex-direction: column;
        width: 100%;

        .cell-content {
          &::before {
            content: attr(data-title);
            position: relative;
            display: inline-flex;
            min-width: 100px;
            padding-right: 0;
            font-weight: 600;
            font-size: 16px;
            color: ${props => props.theme.table.label};
          }
        }
      }
    }
  }

  ${props => props.theme.breakpoints.down(768)} {
    .custom-table {
      .masternode-detail {
        grid-template-columns: 1fr 1fr;
      }
    }
  }

  ${props => props.theme.breakpoints.down(480)} {
    .custom-table {
      .masternode-detail {
        grid-template-columns: 1fr;
      }

      .table__row {
        .cell-content {
          &.see-more {
            width: 100%;

            &::before {
              display: inline-flex;
            }
          }
        }
      }

      .address-link {
        max-width: 57vw;
        display: inline-block;
      }
    }
  }
`;
