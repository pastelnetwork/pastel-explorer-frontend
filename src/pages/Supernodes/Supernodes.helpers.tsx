import { translateDropdown } from '@utils/helpers/i18n';
import { formattedDate } from '@utils/helpers/date/date';
import { INetworkSupernodes } from '@utils/types/INetwork';

import * as Styles from './Supernodes.styles';

export const DATA_FETCH_LIMIT = 20;
export const DATA_OFFSET = 0;
export const DATA_DEFAULT_SORT = 'desc';
export const STATUS_LIST = [
  {
    value: 'all',
    name: 'pages.supernodes.all',
  },
  {
    value: 'PRE_ENABLED',
    name: 'pages.supernodes.preEnabled',
  },
  {
    value: 'ENABLED',
    name: 'pages.supernodes.enabled',
  },
  {
    value: 'EXPIRED',
    name: 'pages.supernodes.expired',
  },
  {
    value: 'WATCHDOG_EXPIRED',
    name: 'pages.supernodes.watchdogExpired',
  },
  {
    value: 'NEW_START_REQUIRED',
    name: 'pages.supernodes.newStartRequired',
  },
  {
    value: 'UPDATE_REQUIRED',
    name: 'pages.supernodes.updateRequired',
  },
  {
    value: 'POSE_BAN',
    name: 'pages.supernodes.poseBan',
  },
  {
    value: 'OUTPOINT_SPENT',
    name: 'pages.supernodes.outpointSpent',
  },
];

export const generateStatusData = (status: string) => {
  return <Styles.Status className={status.toLowerCase()}>{status}</Styles.Status>;
};

export const getCsvHeaders = () => {
  return [
    { key: 'masternodeRank', label: translateDropdown('pages.supernodes.rank') },
    { key: 'supernodeIP', label: translateDropdown('pages.supernodes.supernodeIP') },
    { key: 'address', label: translateDropdown('pages.supernodes.address') },
    { key: 'status', label: translateDropdown('pages.supernodes.status') },
    { key: 'lastPaid', label: translateDropdown('pages.supernodes.lastPaid') },
    { key: 'lastPaidBlock', label: translateDropdown('pages.supernodes.lastPaidBlock') },
    { key: 'dateTimeLastSeen', label: translateDropdown('pages.supernodes.dateTimeLastSeen') },
    { key: 'activeSeconds', label: translateDropdown('pages.supernodes.activeSeconds') },
    { key: 'protocolVersion', label: translateDropdown('pages.supernodes.protocolVersion') },
    { key: 'country', label: translateDropdown('pages.supernodes.country') },
    { key: 'snPastelIdPubkey', label: translateDropdown('pages.supernodes.snPastelIdPubkey') },
    { key: 'pubkey', label: translateDropdown('pages.supernodes.pubkey') },
    { key: 'extAddress', label: translateDropdown('pages.supernodes.extAddress') },
    { key: 'extP2P', label: translateDropdown('pages.supernodes.extP2P') },
  ];
};

export const getCsvData = (supernodes: INetworkSupernodes[]) => {
  if (!supernodes.length) {
    return [];
  }
  return supernodes.map(item => ({
    masternodeRank: item.masternodeRank < 0 ? translateDropdown('common.na') : item.masternodeRank,
    supernodeIP: `${item.ip}:${item.port}`,
    address: item.address,
    status: item.status,
    lastPaid: formattedDate(item.lastPaidTime, {
      dayName: false,
    }),
    lastPaidBlock: item.lastPaidBlock,
    dateTimeLastSeen: formattedDate(item.dateTimeLastSeen, {
      dayName: false,
    }),
    activeSeconds: item.activeSeconds,
    protocolVersion: item.protocolVersion,
    country: item.country,
    snPastelIdPubkey: item.snPastelIdPubkey,
    pubkey: item.pubkey,
    extAddress: item.extAddress,
    extP2P: item.extP2P,
  }));
};
