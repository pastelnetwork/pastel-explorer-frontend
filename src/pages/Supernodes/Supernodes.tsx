import { useState, useEffect, ChangeEvent, useRef, Fragment } from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import IconButton from '@material-ui/core/IconButton';
import Box from '@material-ui/core/Box';
import Tooltip from '@material-ui/core/Tooltip';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ArrowDropDown from '@material-ui/icons/ArrowDropDown';
import ArrowDropUp from '@material-ui/icons/ArrowDropUp';
import parse from 'html-react-parser';

import RouterLink from '@components/RouterLink/RouterLink';
import ExplorerMap from '@pages/Explorer/ExplorerMap/ExplorerMap';
import CopyButton from '@components/CopyButton/CopyButton';
import SupernodeStatistics from '@pages/Explorer/SupernodeStatistics/SupernodeStatistics';
import { INetworkSupernodes } from '@utils/types/INetwork';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import { formattedTimeElapsed, formatFullDate } from '@utils/helpers/date/date';
import { formatAddress } from '@utils/helpers/format';
import { Dropdown, OptionsProps } from '@components/Dropdown/Dropdown';
import useSupernodes from '@hooks/useSupernodes';
import { translate, translateDropdown } from '@utils/helpers/i18n';
import * as ROUTES from '@utils/constants/routes';
import * as TableStyles from '@components/Table/Table.styles';
import * as PastelIdStyles from '@pages/Details/PastelIdDetails/PastelIdDetails.styles';

import {
  STATUS_LIST,
  getCsvHeaders,
  getCsvData,
  generateStatusData,
  DATA_DEFAULT_SORT,
} from './Supernodes.helpers';
import * as Styles from './Supernodes.styles';

function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  if (b[orderBy] < a[orderBy]) {
    return -1;
  }
  if (b[orderBy] > a[orderBy]) {
    return 1;
  }
  return 0;
}

const Supernodes: React.FC = () => {
  const [order, setOrder] = useState(DATA_DEFAULT_SORT);
  const [orderBy, setOrderBy] = useState('lastPaidTime');
  const downloadRef = useRef(null);
  const { masternodes, isLoading } = useSupernodes();
  const [status, setStatus] = useState<string>(STATUS_LIST[0].value);
  const [originalSupernodes, setOriginalSupernodes] = useState<Array<INetworkSupernodes>>([]);
  const [currentSupernodes, setCurrentSupernodes] = useState<Array<INetworkSupernodes>>([]);
  const [supernodes, setSupernodes] = useState<Array<INetworkSupernodes>>([]);
  const [selected, setSelected] = useState<string[]>([]);

  useEffect(() => {
    if (masternodes?.length) {
      setOriginalSupernodes(masternodes);
      setCurrentSupernodes(masternodes);
      setSupernodes(masternodes);
    }
  }, [isLoading, masternodes]);

  const handleChange = (
    event: ChangeEvent<{
      name?: string | undefined;
      value: unknown;
    }>,
  ) => {
    if (event.target.value) {
      setStatus(event.target.value as string);
      let newSupernodes: INetworkSupernodes[] = [];
      if (event.target.value !== 'all') {
        newSupernodes = originalSupernodes.filter(s => s.status === event.target.value);
        setCurrentSupernodes(originalSupernodes.filter(s => s.status === event.target.value));
      } else {
        newSupernodes = originalSupernodes;
        setCurrentSupernodes(originalSupernodes);
      }

      setSupernodes(newSupernodes);
    }
  };

  const generateStatusOptions = () => {
    const results: OptionsProps[] = [];
    STATUS_LIST.map((item: OptionsProps) => {
      let total = originalSupernodes.filter((i: INetworkSupernodes) => i.status === item.value)
        .length;
      if (item.value === 'all') {
        total = originalSupernodes.length;
      }
      if (total <= 0) {
        return item;
      }

      return results.push({
        value: item.value,
        name: `${translateDropdown(item.name)} (${total})`,
      });
    });

    return results;
  };

  const handleToggle = (address: string) => {
    const newList = [...selected];
    if (selected.includes(address)) {
      const index = newList.indexOf(address);
      if (index > -1) {
        newList.splice(index, 1);
      }
    } else {
      newList.push(address);
    }
    setSelected(newList);
  };

  const compare = (a: INetworkSupernodes, b: INetworkSupernodes) => {
    return order === DATA_DEFAULT_SORT
      ? descendingComparator(a, b, orderBy as keyof INetworkSupernodes)
      : -descendingComparator(a, b, orderBy as keyof INetworkSupernodes);
  };

  const handleRequestSort = (property: string) => {
    const isAsc = orderBy === property && order === 'asc';
    setOrder(isAsc ? 'desc' : 'asc');
    setOrderBy(property);
  };

  const renderSortIcon = (field: string) => {
    if (field !== orderBy) {
      return null;
    }
    return order === DATA_DEFAULT_SORT ? <ArrowDropDown /> : <ArrowDropUp />;
  };

  return (
    <>
      <Styles.Gird>
        <Styles.ExplorerMapColumn>
          <ExplorerMap hidePeer />
        </Styles.ExplorerMapColumn>
        <Styles.SupernodeColumn>
          <SupernodeStatistics />
        </Styles.SupernodeColumn>
      </Styles.Gird>
      <Styles.BlockWrapper>
        <Styles.GridWrapper item>
          <TableStyles.BlockWrapper className="mb-0">
            <PastelIdStyles.BlockWrapper className="ticket-title-wrapper">
              <Styles.TitleWrapper>
                <TableStyles.BlockTitle>
                  <Styles.Title>{parse(translate('pages.supernodes.supernodeList'))}</Styles.Title>{' '}
                  {supernodes.length > 0 ? (
                    <Styles.SubTitle>
                      (
                      {parse(
                        translate('pages.supernodes.totalSupernodes', {
                          total: formatNumber(originalSupernodes.length),
                        }),
                      )}
                      )
                    </Styles.SubTitle>
                  ) : null}
                </TableStyles.BlockTitle>
                <div className="filter-wrapper">
                  <Dropdown
                    value={status}
                    onChange={handleChange}
                    options={generateStatusOptions()}
                    label={translateDropdown('pages.supernodes.supernodeStatus')}
                    classNameWrapper="supernode-status"
                  />
                  <Styles.CSVLinkButton
                    data={getCsvData(currentSupernodes)}
                    filename="supernode_list.csv"
                    headers={getCsvHeaders()}
                    separator=","
                    ref={downloadRef}
                    className="download"
                  >
                    {parse(translate('pages.historicalStatistics.downloadCSV'))}
                  </Styles.CSVLinkButton>
                </div>
              </Styles.TitleWrapper>
            </PastelIdStyles.BlockWrapper>
            <Styles.ContentWrapper>
              <TableStyles.PaperWrapper>
                <TableStyles.TableWrapper className="scroll">
                  <Table className="custom-table activities-table offers">
                    <TableHead className="table__row-header">
                      <TableRow>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('ip')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.supernodeIP'))}
                            {renderSortIcon('ip')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('port')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.port'))}
                            {renderSortIcon('port')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('address')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.address'))}
                            {renderSortIcon('address')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('status')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.status'))}
                            {renderSortIcon('status')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('country')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.country'))}
                            {renderSortIcon('country')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell
                          className="nowrap"
                          onClick={() => handleRequestSort('lastPaidTime')}
                        >
                          <span className="header-cell">
                            {parse(translate('pages.supernodes.lastPaid'))}
                            {renderSortIcon('lastPaidTime')}
                          </span>
                        </TableStyles.TableCell>
                        <TableStyles.TableCell className="nowrap">&nbsp;</TableStyles.TableCell>
                      </TableRow>
                    </TableHead>
                    <TableBody>
                      {supernodes?.sort(compare)?.map(masternode => (
                        <Fragment key={`${masternode.ip}:${masternode.port}`}>
                          <TableRow
                            className={`table__row ${
                              selected.includes(masternode.address) ? 'row_active' : ''
                            }`}
                          >
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.supernodeIP')}:`}
                              className="cell-content nowrap"
                            >
                              {masternode.ip}
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.port')}:`}
                              className="cell-content nowrap"
                            >
                              {masternode.port}
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.address')}:`}
                              className="cell-content nowrap read-more"
                            >
                              <span className="wrapper-content">
                                <CopyButton copyText={masternode.address} />
                                <RouterLink
                                  route={`${ROUTES.ADDRESS_DETAILS}/${masternode.address}`}
                                  value={masternode.address}
                                  textSize="large"
                                  title={masternode.address}
                                  className="address-link"
                                />
                              </span>
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.status')}:`}
                              className="cell-content nowrap"
                            >
                              {generateStatusData(masternode.status)}
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.country')}:`}
                              className="cell-content nowrap"
                            >
                              {masternode.country}
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.lastPaid')}:`}
                              className="cell-content nowrap"
                            >
                              {formattedTimeElapsed(masternode.lastPaidTime)}
                            </TableStyles.RowCell>
                            <TableStyles.RowCell
                              data-title={`${translateDropdown('pages.supernodes.seeMore')}:`}
                              className="cell-content nowrap see-more"
                              width="50px"
                            >
                              <Tooltip
                                title={
                                  selected.includes(masternode.address)
                                    ? translateDropdown('pages.supernodes.seeLess')
                                    : translateDropdown('pages.supernodes.seeMore')
                                }
                              >
                                <IconButton
                                  onClick={() => handleToggle(masternode.address)}
                                  className={selected.includes(masternode.address) ? 'open' : ''}
                                >
                                  <ExpandMoreIcon />
                                </IconButton>
                              </Tooltip>
                            </TableStyles.RowCell>
                          </TableRow>
                          {selected.includes(masternode.address) ? (
                            <TableRow className="row_more_active">
                              <TableStyles.TableCell colSpan={7}>
                                <div className="masternode-detail">
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.rank'))}
                                    </Box>
                                    <Box className="bold">
                                      {masternode.masternodeRank < 0
                                        ? parse(translate('common.na'))
                                        : masternode.masternodeRank}
                                    </Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.lastPaidBlock'))}
                                    </Box>
                                    <Box className="bold">
                                      <RouterLink
                                        route={`${ROUTES.BLOCK_DETAILS}/${masternode.lastPaidBlock}`}
                                        value={masternode.lastPaidBlock}
                                        textSize="large"
                                        title={masternode.lastPaidBlock.toString()}
                                        className="address-link"
                                      />
                                    </Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.snPastelIdPubkey'))}
                                    </Box>
                                    <Box className="bold">
                                      <span className="wrapper-content">
                                        <CopyButton copyText={masternode.snPastelIdPubkey} />
                                        <RouterLink
                                          route={`${ROUTES.PASTEL_ID_DETAILS}/${masternode.snPastelIdPubkey}`}
                                          value={formatAddress(masternode.snPastelIdPubkey, 5, -5)}
                                          textSize="large"
                                          title={masternode.snPastelIdPubkey}
                                          className="address-link"
                                        />
                                      </span>
                                    </Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.city'))}
                                    </Box>
                                    <Box className="bold">{masternode.city}</Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.rankAsOfBlockHeight'))}
                                    </Box>
                                    <Box className="bold">
                                      {masternode.rankAsOfBlockHeight < 0 ? (
                                        parse(translate('common.na'))
                                      ) : (
                                        <RouterLink
                                          route={`${ROUTES.BLOCK_DETAILS}/${masternode.rankAsOfBlockHeight}`}
                                          value={masternode.rankAsOfBlockHeight}
                                          textSize="large"
                                          title={masternode.rankAsOfBlockHeight.toString()}
                                          className="address-link"
                                        />
                                      )}
                                    </Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.activeSeconds'))}
                                    </Box>
                                    <Box className="bold">{masternode.activeSeconds}</Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.protocolVersion'))}
                                    </Box>
                                    <Box className="bold">{masternode.protocolVersion}</Box>
                                  </div>
                                  <div>
                                    <Box className="title">
                                      {parse(translate('pages.supernodes.dateTimeLastSeen'))}
                                    </Box>
                                    <Box className="bold">
                                      {formatFullDate(masternode.dateTimeLastSeen * 1000)}
                                    </Box>
                                  </div>
                                </div>
                              </TableStyles.TableCell>
                            </TableRow>
                          ) : null}
                        </Fragment>
                      ))}
                    </TableBody>
                  </Table>
                </TableStyles.TableWrapper>
              </TableStyles.PaperWrapper>
            </Styles.ContentWrapper>
          </TableStyles.BlockWrapper>
        </Styles.GridWrapper>
      </Styles.BlockWrapper>
    </>
  );
};

export default Supernodes;
