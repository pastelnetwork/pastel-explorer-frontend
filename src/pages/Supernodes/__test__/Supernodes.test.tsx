import { shallow } from 'enzyme';

import 'jest-styled-components';

import i18next from '../../../utils/helpers/i18n';
import Supernodes from '../Supernodes';

jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
});

describe('pages/Supernodes', () => {
  const wrapper = shallow(<Supernodes />);

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
