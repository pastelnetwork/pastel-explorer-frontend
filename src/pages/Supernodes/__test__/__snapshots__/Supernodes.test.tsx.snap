// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Supernodes renders correctly 1`] = `ShallowWrapper {}`;
